### ISSUE TITLE
Error: Bad packet length thrown from SSH2Stream._transform

### ISSUE THREAD

**Issue by makeusabrew** [mscdex/ssh2-streams#9]
Hey Brian,

I'm running into this error occasionally when forwarding data through an `accept` stream returned from the `ssh2` module's `client.on("tcp connection", ...)` call. You'll notice that some of the debug below comes from my module but hopefully it doesn't obscure anything. The scenario below is the result of a uploading a ~935Kb file (I use ssh2 to set up a reverse proxy) - you can see the initial POST kicks things off, and every `REMOTE --> LOCAL` line is emitted when the reading stream receives data:

```
  finch:core:tunnel Issuing request: POST / +1ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETBEFORE (expecting 16) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKET +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: pktLen:2924,padLen:18,remainLen:2912 +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATA +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: HMAC size:16 +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY +1ms
  finch:core:tunnel DEBUG: Parser: Verifying MAC +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (2) +0ms
  finch:core:tunnel REMOTE --> LOCAL +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETBEFORE (expecting 16) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKET +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: pktLen:2924,padLen:18,remainLen:2912 +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATA +1ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: HMAC size:16 +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY +0ms
  finch:core:tunnel DEBUG: Parser: Verifying MAC +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (2) +0ms
  finch:core:tunnel REMOTE --> LOCAL +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETBEFORE (expecting 16) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKET +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: pktLen:2924,padLen:18,remainLen:2912 +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATA +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: HMAC size:16 +1ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY +0ms
  finch:core:tunnel DEBUG: Parser: Verifying MAC +1ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (2) +1ms
  finch:core:tunnel REMOTE --> LOCAL +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETBEFORE (expecting 16) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKET +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: pktLen:1468,padLen:10,remainLen:1456 +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATA +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: HMAC size:16 +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY +0ms
  finch:core:tunnel DEBUG: Parser: Verifying MAC +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (2) +0ms
  finch:core:tunnel REMOTE --> LOCAL +1ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETBEFORE (expecting 16) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKET +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: pktLen:2924,padLen:18,remainLen:2912 +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATA +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: HMAC size:16 +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY +0ms
  finch:core:tunnel DEBUG: Parser: Verifying MAC +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (2) +1ms
  finch:core:tunnel REMOTE --> LOCAL +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETBEFORE (expecting 16) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKET +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: pktLen:2924,padLen:18,remainLen:2912 +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATA +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Parser: HMAC size:16 +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY +0ms
  finch:core:tunnel DEBUG: Parser: Verifying MAC +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (2) +1ms
  finch:core:tunnel REMOTE --> LOCAL +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKETBEFORE (expecting 16) +0ms
  finch:core:tunnel DEBUG: Parser: IN_PACKET +0ms
  finch:core:tunnel DEBUG: Parser: Decrypting +0ms
  finch:core:tunnel DEBUG: Outgoing: Writing DISCONNECT (PROTOCOL_ERROR) +0ms
  finch:core:tunnel DEBUG: Parser: Bad packet length (65584) +0ms

Error: Bad packet length
  at SSH2Stream._transform (/home/nick/code/node/finch/client/core/node_modules/ssh2/node_modules/ssh2-streams/lib/ssh.js:412:25)
  at SSH2Stream.Transform._read [as __read] (_stream_transform.js:179:10)
  at SSH2Stream._read (/home/nick/code/node/finch/client/core/node_modules/ssh2/node_modules/ssh2-streams/lib/ssh.js:213:15)
  at SSH2Stream.Transform._write (_stream_transform.js:167:12)
  at doWrite (_stream_writable.js:226:10)
  at writeOrBuffer (_stream_writable.js:216:5)
  at SSH2Stream.Writable.write (_stream_writable.js:183:11)
  at write (_stream_readable.js:582:24)
  at flow (_stream_readable.js:591:7)
  at Socket.pipeOnReadable (_stream_readable.js:623:5)
  at Socket.EventEmitter.emit (events.js:92:17)
  at emitReadable_ (_stream_readable.js:407:10)
  at emitReadable (_stream_readable.js:403:5)
  at readableAddChunk (_stream_readable.js:165:9)
  at Socket.Readable.push (_stream_readable.js:127:10)
  at TCP.onread (net.js:528:21)
```

I can supply a source code snippet if it helps but all that's happening is the `accept` stream is being piped into a local `net.Socket`, and that socket is being piped back into the accept stream. The code in my wrapper module hasn't changed much for a year or so, and indeed while debugging things today I swapped back to ssh2 0.3.3 (i.e. pre rewrite / ssh2-streams split) and the same scenario and file works reliably every time.

I must admit the inner workings of the ssh2-streams module are a bit beyond my limited knowledge of the protocol :). What I have found is that each stream's maxPacketSize is (effectively) hard-coded to 35000 bytes; although it is an optional constructor param of each stream, users of the ssh2 module can't configure the value which is passed in:

https://github.com/mscdex/ssh2-streams/blob/master/lib/ssh.js#L104

Initially I noticed that the bad packet was suspiciously close to the max TCP packet size of 64Kb (albeit with some padding which I put down to SSH2 protocol padding), so manually doubled MAX_PACKET_SIZE to 70000. That alone wasn't enough due to:

https://github.com/mscdex/ssh2-streams/blob/master/lib/ssh.js#L3159-L3162

Doubling that figure in tandem with a larger max packet size solved the issue. Just to humour myself I tried MAX_CHAN_DATA_LEN again with a value of (32768*2)-1 and sure enough tripped the limit again.

As you can tell, I don't know enough to pose the correct solution but wanted to outline the problem with as much information as I could. For what it's worth, and although I can't repeat it again now, I also saw the issue occur with a much larger final packet of around 750Kb after several 16Kb packets. I realise without debug output that's anecdotal, and has only happened maybe twice out of 100 or so attempts, but it *did* happen (honest). I manually tallied the total packet sizes in that instance and they matched up to the size of the file being uploaded, so I'm fairly confident what I saw was correct. After a bit more digging I do believe that's perfectly possible, since a `data` event isn't necessarily coupled 1:1 with an underlying TCP packet.

One thing which might be relevant (or not): the call to `TransformStream.push` with the big packet returns `false`, i.e. the stream is starting to be buffered at this point. I know one can continue to write even in such a situation, but just trying to supply as much information as possible.

Anything else you need or want me to debug more specifically, please do just let me know.

All the best, and thanks again for your great work :)

Nick

**Comment by mscdex** [comment]
If you apply [this patch](https://gist.github.com/mscdex/3695e2ecd0fa860d7850) to `ssh2` does it solve the problem?

**Comment by makeusabrew** [comment]
@mscdex you're a star. Not quite on its own, but also changing the same variable in the `self._sshstream.channelOpenConfirm` call a few lines later (https://github.com/mscdex/ssh2/blob/master/lib/client.js#L1292) seems to have fixed things. Thank you so much :)

**Comment by mscdex** [comment]
Fixed in mscdex/ssh2@0183bae8d4. Thanks for the report.
